tests/test_FLEbasis2D.py

```diff
@@ -0,0 +1,330 @@
+    for i in range(4):
+        bandlimit = L // (2**i)
+        coeffs_lowpassed = basis.lowpass(coeffs, bandlimit)
+        nonzero_coeffs.append(np.sum(coeffs_lowpassed != 0))
+
+    # for bandlimit == L, no frequencies should be removed
+    assert nonzero_coeffs[0] == basis.count
+
+    # for lower bandlimits, there should be fewer and fewer nonzero coeffs
+    assert nonzero_coeffs[0] > nonzero_coeffs[1] > nonzero_coeffs[2] > nonzero_coeffs[3]
+
+    # make sure you can pass in a 1-D array if you want
+    _ = basis.lowpass(coeffs[0, :], L)
+
+    # cannot pass in the wrong number of coefficients
+    with pytest.raises(
+        AssertionError, match="Number of coefficients must match self.count."
+    ):
+        _ = basis.lowpass(coeffs[:, :1000], L)
+
+    # cannot pass in wrong shape
+    with pytest.raises(
+        AssertionError,
+        match="Input a stack of coefficients of dimension",
+    ):
+        _ = basis.lowpass(np.zeros((3, 3, 3)), L)
+
+
+def testRotate():
+    # test ability to accurately rotate images via
+    # FLE coefficients
+
+    L = 128
+    basis = FLEBasis2D(L, match_fb=False)
+
+    # sample image
+    ims = create_images(L, 1)
+    # rotate 90 degrees in cartesian coordinates
+    ims_90 = Image(np.rot90(ims.asnumpy(), axes=(1, 2)))
+
+    # get FLE coefficients
+    coeffs = basis.evaluate_t(ims)
+    coeffs_cart_rot = basis.evaluate_t(ims_90)
+
+    # rotate original image in FLE space
+    coeffs_fle_rot = basis.rotate(coeffs, np.pi / 2)
+
+    # back to cartesian
+    ims_cart_rot = basis.evaluate(coeffs_cart_rot)
+    ims_fle_rot = basis.evaluate(coeffs_fle_rot)
+
+    # test rot90 close
+    assert np.allclose(ims_cart_rot[0], ims_fle_rot[0], atol=1e-4)
+
+    # 2Pi identity in FLE space (rotate by 2Pi)
+    coeffs_fle_2pi = basis.rotate(coeffs, 2 * np.pi)
+    ims_fle_2pi = basis.evaluate(coeffs_fle_2pi)
+
+    # test 2Pi identity
+    assert np.allclose(ims[0], ims_fle_2pi[0], atol=utest_tolerance(basis.dtype))
+
+    # Reflect in FLE space (rotate by Pi)
+    coeffs_fle_pi = basis.rotate(coeffs, np.pi)
+    ims_fle_pi = basis.evaluate(coeffs_fle_pi)
+
+    # test reflection
+    assert np.allclose(np.flipud(ims[0]), ims_fle_pi[0], atol=1e-4)
+
+    # make sure you can pass in a 1-D array if you want
+    _ = basis.lowpass(np.zeros((basis.count,)), np.pi)
+
+    # cannot pass in the wrong number of coefficients
+    with pytest.raises(
+        AssertionError, match="Number of coefficients must match self.count."
+    ):
+        _ = basis.rotate(np.zeros((1, 10)), np.pi)
+
+    # cannot pass in wrong shape
+    with pytest.raises(
+        AssertionError,
+        match="Input a stack of coefficients of dimension",
+    ):
+        _ = basis.lowpass(np.zeros((3, 3, 3)), np.pi)
+
+
+def testRadialConvolution():
+    # test ability to accurately convolve with a radial
+    # (e.g. CTF) function via FLE coefficients
+
+    L = 32
+    basis = FLEBasis2D(L, match_fb=False)
+    # load test CTF
```

Comment:
https://github.com/ComputationalCryoEM/ASPIRE-Python/issues/738#issuecomment-1428703187

Reply:
Looking... I'm wondering if what they provided is actually a ctf? 

I'm pretty comfortable saying our CTFFilter is making something reasonable (already in F space), and some code to use (convolve) it manually is in Image.filter. The code is not radically different than the PR example I pointed you towards.. or what you mention in your write up.

There are known issues of padding/masking, but these are just improvements, not show stoppers usually.  However,  it looks like the function defined by that array is radial in real space (only)? or just very weak in second dim. Unsure.  The real function looks much more like a CTF then the transformed one....

Because this fast ctf thing is one of the main reasons to use FLE, this is moderately concerning... at least from what I understand atm.  Maybe we can ask Oscar or Nick how exactly they generated this, and what the expectations are.

I understand that you may not be able to resolve this, but I think we may want to consider not labeling test/vars/data as CTF... especially if you are confident you can prove that using a CTF is _not_ working...

Thanks for documenting.

Reply:
Sounds good, renamed vars, etc to be a more generic radial function